stan/math/prim/scal/prob/normal_lcdf.hpp

```diff
@@ -53,34 +55,132 @@ inline return_type_t<T_y, T_loc, T_scale> normal_lcdf(const T_y& y,
-      one_p_erf = 0.0;
-    } else if (scaled_diff < -5.0 * INV_SQRT_2) {
-      one_p_erf = erfc(-scaled_diff);
-    } else if (scaled_diff > 8.25 * INV_SQRT_2) {
-      one_p_erf = 2.0;
+    const T_partials_return sigma_sqrt2 = sigma_dbl * SQRT_2;
+    const T_partials_return x2 = square(scaled_diff);
+
+    // check whether scaled_diff^10 term will overflow
+    const bool overflow_check
+        = 10.0 * log(-scaled_diff)
+          < log(std::numeric_limits<T_partials_return>::max());
+
+    // use erfc() instead of erf() in order to retain precision
+    // since for x>0 erfc()->0
+    if (scaled_diff > 0.0) {
+      // CDF(x) = 1/2 + 1/2erf(x) = 1 - 1/2erfc(x)
+      cdf_log += log1p(-0.5 * erfc(scaled_diff));
+      if (isnan(cdf_log)) {
+        cdf_log = 0;
+      }
+    } else if (scaled_diff > -20.0) {
+      // CDF(x) = 1/2 - 1/2erf(-x) = 1/2erfc(-x)
+      cdf_log += log(erfc(-scaled_diff)) + LOG_HALF;
     } else {
-      one_p_erf = 1.0 + erf(scaled_diff);
+      if (overflow_check) {
+        // entering territory where erfc(-x)~0
+        // need to use direct numerical approximation of cdf_log instead
+        // the following based on W. J. Cody, Math. Comp. 23(107):631-638 (1969)
+        // CDF(x) = 1/2erfc(-x)
+        const T_partials_return x4 = pow(scaled_diff, 4);
+        const T_partials_return x6 = pow(scaled_diff, 6);
+        const T_partials_return x8 = pow(scaled_diff, 8);
+        const T_partials_return x10 = pow(scaled_diff, 10);
+        const T_partials_return temp_p
+            = 0.000658749161529837803157 + 0.0160837851487422766278 / x2
+              + 0.125781726111229246204 / x4 + 0.360344899949804439429 / x6
+              + 0.305326634961232344035 / x8 + 0.0163153871373020978498 / x10;
+        const T_partials_return temp_q
+            = -0.00233520497626869185443 - 0.0605183413124413191178 / x2
+              - 0.527905102951428412248 / x4 - 1.87295284992346047209 / x6
+              - 2.56852019228982242072 / x8 - 1.0 / x10;
+        cdf_log += LOG_HALF + log(1.0 / SQRT_PI + (temp_p / temp_q) / x2)
+                   - log(-scaled_diff) - x2;
+      } else {
+        cdf_log = stan::math::negative_infinity();
+      }
     }
 
-    cdf_log += LOG_HALF + log(one_p_erf);
-
     if (!is_constant_all<T_y, T_loc, T_scale>::value) {
-      const T_partials_return rep_deriv_div_sigma
-          = scaled_diff < -37.5 * INV_SQRT_2
-                ? std::numeric_limits<double>::infinity()
-                : SQRT_TWO_OVER_PI * exp(-scaled_diff * scaled_diff) / sigma_dbl
-                      / one_p_erf;
+      // compute partial derivatives
+      // based on analytic form given by:
+      // dln(CDF)/dx = exp(-x^2)/(sqrt(pi)*(1/2+erf(x)/2)
+      T_partials_return dncdf_log_dbl = 0.0;
+      T_partials_return t = 0.0, t2 = 0.0, t4 = 0.0;
+
+      // calculate using piecewise funciton
+      // (due to instability / inaccuracy in the various approximations)
+      if (scaled_diff < 0.1) {
```

Comment:
In the same way as above would it be possible to flatten these conditionals on the piece-wise defined derivatives? It looks like they're all defined in terms of scaled_diff.

Also would it be possible to write them in order of what interval they represent using similar notation? Like it's difficult to parse sometimes scaled_diff > 1.5 sometimes scaled_diff < 2.2 or whatever. If it's necessary to write it that way it's fine just checking :D.

Reply:
Yep that makes a lot more sense. I think the way to do this would be

`if (scaled_diff > 2.2) {`
`...`
`} else if (scaled_diff > 1.5) {`
`...`
`} else if (scaled_diff > 0.8) {`
`...`
`} else if (scaled_diff > 0.1) {`
`...`
`} else if (scaled_diff > -4.0) {`
`...`
`} else {`
`...`
`}`

although there will be some code duplication in the 4th and 5th statements (but I'm guessing this is ok if it improves code readability?).